core/modules/flowgen.cc

```diff
@@ -405,26 +435,67 @@ void FlowGen::DeInit() {
     flows_free_.pop();
   }
 
-  delete[] templ_;
+  delete[] tmpl_;
 }
 
 CommandResponse FlowGen::UpdateBaseAddresses() {
-  char *p = reinterpret_cast<char *>(templ_);
+  char *p = reinterpret_cast<char *>(tmpl_);
   if (!p) {
     return CommandFailure(EINVAL, "must specify 'template'");
   }
 
   Ipv4 *ipheader = reinterpret_cast<Ipv4 *>(p + sizeof(Ethernet));
-  Tcp *tcpheader = reinterpret_cast<Tcp *>(p + sizeof(Ethernet) + sizeof(Ipv4));
-
   ip_src_base_ = ipheader->src.value();
   ip_dst_base_ = ipheader->dst.value();
-  port_src_base_ = tcpheader->src_port.value();
-  port_dst_base_ = tcpheader->dst_port.value();
+
+  if (l4_proto_ == Ipv4::Proto::kUdp) {
```

Comment:
Most existing code of other modules (e.g., ipswap or nat) does not differentiate TCP and UDP when only L4 ports are concerned, to reduce code redundancy. UDP and TCP share the same layout for src/dst ports for this purpose.

```
Udp *l4_header = reinterpret_cast<Udp *>(p + sizeof(Ethernet) + sizeof(Ipv4));
port_src_base_ = l4_header->src_port.value();
port_dst_base_ = l4_header->dst_port.value();
```
should be enough for both UDP and TCP. You may also want to declare `using L4 = Udp;` somewhere to feel less guilty...

Reply:
Sad :(